src/pkcs15init/pkcs15-lib.c

```diff
@@ -2625,32 +2626,38 @@ prkey_fixup_rsa(struct sc_pkcs15_card *p15card, struct sc_pkcs15_prkey_rsa *key)
 	 /* We don't really need an RSA structure, only the BIGNUMs */
 
 	if (!key->dmp1.len || !key->dmq1.len || !key->iqmp.len) {
-		BIGNUM *aux;
-		BN_CTX *bn_ctx;
-		BIGNUM *rsa_n, *rsa_e, *rsa_d, *rsa_p, *rsa_q, *rsa_dmp1, *rsa_dmq1, *rsa_iqmp;
-
-		rsa_n = BN_bin2bn(key->modulus.data, key->modulus.len, NULL);
-		rsa_e = BN_bin2bn(key->exponent.data, key->exponent.len, NULL);
-		rsa_d = BN_bin2bn(key->d.data, key->d.len, NULL);
-		rsa_p = BN_bin2bn(key->p.data, key->p.len, NULL);
-		rsa_q = BN_bin2bn(key->q.data, key->q.len, NULL);
-		rsa_dmp1 = BN_new();
-		rsa_dmq1 = BN_new();
-		rsa_iqmp = BN_new();
-
-		aux = BN_new();
-		bn_ctx = BN_CTX_new();
-
-		BN_sub(aux, rsa_q, BN_value_one());
-		BN_mod(rsa_dmq1, rsa_d, aux, bn_ctx);
-
-		BN_sub(aux, rsa_p, BN_value_one());
-		BN_mod(rsa_dmp1, rsa_d, aux, bn_ctx);
+		BIGNUM *aux = NULL;
+		BN_CTX *bn_ctx = NULL;
+		BIGNUM *rsa_n = NULL, *rsa_e = NULL, *rsa_d = NULL, *rsa_p = NULL,
+			*rsa_q = NULL, *rsa_dmp1 = NULL, *rsa_dmq1 = NULL, *rsa_iqmp = NULL;
+
+		if (!(rsa_n = BN_bin2bn(key->modulus.data, key->modulus.len, NULL)) ||
+				!(rsa_e = BN_bin2bn(key->exponent.data, key->exponent.len, NULL)) ||
+				!(rsa_d = BN_bin2bn(key->d.data, key->d.len, NULL)) ||
+				!(rsa_p = BN_bin2bn(key->p.data, key->p.len, NULL)) ||
+				!(rsa_q = BN_bin2bn(key->q.data, key->q.len, NULL)) ||
+				!(rsa_q = BN_bin2bn(key->q.data, key->q.len, NULL))) {
+			r = SC_ERROR_INTERNAL;
+			goto end;
+		}
 
-		BN_mod_inverse(rsa_iqmp, rsa_q, rsa_p, bn_ctx);
+		if (!(rsa_dmp1 = BN_new()) ||
+				!(rsa_dmq1 = BN_new()) ||
+				!(rsa_iqmp = BN_new()) ||
+				!(aux = BN_new()) ||
+				!(bn_ctx = BN_CTX_new())) {
```

Comment:
nit: I would rather see the assignments outside of the `if` condition and then separate if condition checking the results instead of fitting everything into the multiline `if` condition as these results are not dependent on each other (unlike the following part where the calculation is already done).